proposed/2023/NuGetAudit-without-nuget.org-as-packageSource.md

```diff
@@ -0,0 +1,190 @@
+
+## Motivation
+
+<!-- Why are we doing this? What pain points does this solve? What is the expected outcome? -->
+[NuGetAudit](https://learn.microsoft.com/nuget/concepts/auditing-packages) is a feature where NuGet will report during restore which packages with known vulnerabilities are being used by projects, [proposed in late 2022](../2022/vulnerabilities-in-restore.md), and finally released with .NET 8 in 2023.
+The first version of NuGetAudit used package sources as the data source for known vulnerabilities, which will be referred to as a Vulnerability Database (VDB) for the remainder of this document.
+While nuget.org provides a VDB, there are several reasons why customers may not be using nuget.org as a package source.
+This is especially true following the 2021 blog post titled "Dependency confusion" which significantly increased knowledge of multi-source package substitution supply chain risks.
+
+While the design of NuGetAudit attempts to make it easy for upstreaming package sources to also upstream the `VulnerabilityInfo` resource, the NuGet team and customers are reliant of the feed implementor to implement the change.
+An upstreaming package source is one that either aggregates multiple package sources or caches packages from another package source.
+Upstreaming isn't a concept that exists in NuGet, so servers have to present to clients as if all data (packages, search results, vulnerability information) originates from itself.
+
+However, there are also cases where customers are not using upstreaming feeds, but instead curate approved 3rd party packages and manually upload them to an internal feed.
+There are also other scenarios where customers do not use nuget.org (directly) as a package source.
+
+It's possible for customers to use [Package Source Mapping](https://learn.microsoft.com/nuget/consume-packages/package-source-mapping) to add nuget.org as a package source, but configured in a way that NuGet will never restore packages from this package source.
+However, some developers might consider this an unacceptable risk and prefer another way to get a VDB that cannot accidentally be used to get packages.
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+*NuGet.Config* files have a new section `<auditSources>` where the URL(s) of any audit sources can be specified, which will be used only to download a VDB, and will not be used for downloading packages.
+Like `<packageSources>`, `<auditSources>` can contain zero, one, or more sources, and supports `<clear />`.
+When there is at least one source in `auditSources`, NuGet should no longer use `packageSources` to get a VDB.
```

Comment:
My interpretation without documentation for this keyword would be that all `packageSources `are `auditSources `by default and any source mentioned as `auditSources` would be additional audit source unless `</clear>` is used which will obviously clear all packageSources as audit sources

Reply:
That's an interesting interpretation, but not what I intended.

As you know, nuget accumulates settings from multiple nuget.config files. Therefore, `<clear />` within an `<auditSources>` means "if a lower priority nuget.config file has already defined any audit source, remove it". This is important in CI scenarios where a "stateful" agent is used, since another build on the same agent might have modified the user-profile nuget.config file.

Once NuGet has generated the effective settings from all nuget.config files, NuGet now has a list of packageSources and potentially a list of audit sources. If audit sources list length is zero, then package sources are used as audit sources. Otherwise, only audit sources are used to download a vulnerabilities database.

Reply:
another perspective here that i'm seeing from anand's feedback is that yes we absolutely imply a package source is an audit source, but the specific use-case we are trying to address here is when you don't have a package source that you want to use for auditing data. 

so in all, if you specified nuget.org as a package source, you wouldn't need to imply it as an audit source as it would be implicit.

but if you say run only on an internal source, you may absolutely want to use the nuget.org audit source. this is what i meant earlier by advanced scenario. 

the core scenario would be that people are using package sources that have auditing capabilities enabled. but the adv scenario is they override it using this wonderful spec of `<auditSources>`.

Reply:
I don't understand what's actionable on my side. Are you (both?) proposing that I remove the line that this comment thread is linked to (don't use package sources as audit sources, when there's at least one audit source explicitly defined). Are you saying you want the spec changed so that `auditSources` is **always** additive to the list of `packageSources`?

My position is that it was a mistake to use package sources as audit sources in the first place. Overloading a single thing for multiple reasons just generally complicates things. I don't think that having customers duplicate feed URLs (same URL in both `packageSources` and `nugetSources` sections) is a problem. However, completely stopping using package sources as audit sources would be a breaking change from NuGet 6.8. Hence why this spec proposes using packageSources for NuGetAudit only when zero `auditSources` are defined.

While it can be argued that using package sources for NuGetAudit makes is "easier" for customers to use, it has the unintended consequence that authenticated feeds that don't provide vulnerability info may cause NuGet to prompt interactively for credentials when there's no benefit (no packages needed from the feed). Perhaps some people will think this is a small edge case that doesn't matter, but I don't agree, and the point is that it's not a "zero downside" design choice, or even "obviously better, noone will disagree" design choice.

Reply:
The actionable part here I believe is that we should ensure that any defined packageSource that supports VDB is implicitly added to the auditSources (doesn't have to be written to file).

The customer should be able to include a reputable package source in their packageSources and for features to "just work". When a customer has an advanced use-case, they use the auditSources overriding behavior here.